### PR TITLE
METRON-2023 [UI] Covering Grok Parser Creation with Cypress tests

### PR DESCRIPTION
**Pull request by tiborm**
## Contributor Comments

In this PR I introducing Cypress in Config UI and adding a Cypress test to the Config UI to cover Grok Parser Creation. To do this I made a simplified common test setup across Config and Alert UI to bring them to the same platform.
I also had to remove maven-frontend-plugin to fix issues we faced with on Travis while we tried to add cypress to Config UI project.

The list of changes are the following:
- adding Grok Parser creation test
- removing additional test server solution from Alert UI
- using standard ng serve command to fire up both Alert and Config UI's for testing 
- removing maven-frontend-plugin, excluding UI from maven install and test loops and letting tests run on Travis's own node.js (the issue was deeply investigated and this is the only way to fix issues in the cypress install phase)
- introducing shell script to execute UI tests

## Steps to validate

Please check build logs in Travis CI to see the output of the UI tests.
https://travis-ci.org/apache/metron/jobs/502040547

Under [this link](https://travis-ci.org/tiborm/metron/jobs/502000711) you can find a build failing where I intentionally broke a test to see whether it breaks the build too.

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel).
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```

- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -23,7 +23,7 @@ env:
   - SCRIPT="mvn surefire:test@unit-tests -T 2C"
   - SCRIPT="mvn surefire:test@integration-tests"
   - SCRIPT="mvn clean site --projects site-book"
-  - SCRIPT="mvn test --projects metron-interface/metron-config,metron-interface/metron-alerts"
+  - SCRIPT="./metron-interface/scripts/run-all-ui-test.sh"
```
Comment (tiborm) [inline .travis.yml]: Migrated to bash script from maven. manven-frontend-plugin thrown error when tried to unzip Cypress.

File: .travis.yml [hunk 2]
```diff
@@ -53,8 +51,5 @@ before_cache:
 cache:
   timeout: 1000
   directories:
-  - $HOME/.npm-cache
-  - $HOME/.npm-prefix
-  - metron-interface/metron-config/node_modules
```
Comment (tiborm) [inline .travis.yml]: Migrated to Travis CI's own node/npm instance managed by nvm. The above configs no longer needed.

File: metron-interface/metron-alerts/.nvmrc [hunk 3]
```diff
@@ -1 +1 @@
-v9.11.1
+v10.15.1
```
Comment (tiborm) [inline metron-interface/metron-alerts/.nvmrc]: With .nvmrc we can pin down the node/npm version both on Travis and local.
Upgraded to the latest LTS node version.

File: metron-interface/metron-alerts/package.json [hunk 4]
```diff
@@ -68,15 +65,14 @@
...
     "serve-favicon": "^2.5.0",
-    "serve-static": "^1.13.2",
     "ssh2": "^0.5.5",
+    "start-server-and-test": "^1.7.11",
+    "ts-node": "^7.0.0",
     "tslint": "~5.0.0",
-    "typescript": "~2.9.2",
-    "yamljs": "^0.3.0"
+    "typescript": "~2.9.2"
   }
```
Comment (tiborm) [inline metron-interface/metron-alerts/package.json]: With the simplification of the way we fire up UIs for testing in Travis, we're able to eliminate 5 of our dependencies with adding only one.

File: metron-interface/metron-alerts/pom.xml [hunk 5]
```diff
@@ -22,122 +22,4 @@
...
-                        </goals>
-                        <configuration>
-                            <executable>./scripts/prepend_license_header.sh</executable>
-                        </configuration>
-                    </execution>
-                </executions>
-            </plugin>
-        </plugins>
-    </build>
 </project>
```
Comment (tiborm) [inline metron-interface/metron-alerts/pom.xml]: I kept the pom.xml to have our two UI's in the maven project tree (makes it easier to import etc.), but I removed the two UI from the install and test goals.
Also tried to upgrade to the latest maven-frontend-plugin 1.7.5 but it hasn't fixed the unzipping issue.

File: metron-interface/metron-alerts/scripts/run-all-test.sh [hunk 6]
```diff
@@ -14,10 +15,21 @@
...
+
+npm run test:ci
+npm run cypress:ci
+
+# licence check needs an initial build step
+npm run build
+echo "┌─────────────────────────────────────┐"
+echo "| Checking Apache license headers...  |"
+echo "└─────────────────────────────────────┘"
+./scripts/prepend_license_header.sh
```
Comment (tiborm) [inline metron-interface/metron-alerts/scripts/run-all-test.sh]: For some reason, git recognized this change as a rename. It is actually a removal of a yml config file (belongs to our previous test server setup, no longer needed) and the addition of a shell scirpt.

File: metron-interface/metron-config/cypress/fixtures/config.json [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+              ],
+              "transformation": "IP_PROTOCOL",
+              "config": {}
+          }
+      ],
+      "cacheConfig": {},
+      "rawMessageStrategy": "DEFAULT",
+      "rawMessageStrategyConfig": {}
+  }
+}
```
Comment (tiborm) [inline metron-interface/metron-config/cypress/fixtures/config.json]: Response mock for GROK parser creation test.

File: metron-interface/metron-config/cypress/fixtures/sensor-config-single-parser/config-list-available.json [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+  "FireEye": "org.apache.metron.parsers.fireeye.BasicFireEyeParser",
+  "Asa": "org.apache.metron.parsers.asa.BasicAsaParser",
+  "Syslog3164": "org.apache.metron.parsers.syslog.Syslog3164Parser",
+  "BaseSyslog": "org.apache.metron.parsers.syslog.BaseSyslogParser",
+  "Snort": "org.apache.metron.parsers.snort.BasicSnortParser",
+  "JSONMap": "org.apache.metron.parsers.json.JSONMapParser",
+  "Ise": "org.apache.metron.parsers.ise.BasicIseParser",
+  "GrokWebSphere": "org.apache.metron.parsers.websphere.GrokWebSphereParser",
+  "Bro": "org.apache.metron.parsers.bro.BasicBroParser"
+}
```
Comment (tiborm) [inline metron-interface/metron-config/cypress/fixtures/sensor-config-single-parser/config-list-available.json]: Response mock for GROK parser creation test.

File: metron-interface/metron-config/scripts/start-server-for-e2e.sh [hunk 9]
```diff
@@ -14,10 +15,5 @@
...
-port: 4201
-
-rest:
-  host: localhost
-  port: 8080
-
-dirPath: alerts-ui
\ No newline at end of file
+SCRIPTS_ROOT="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+$SCRIPTS_ROOT/../node_modules/@angular/cli/bin/ng serve --proxy-config proxy.conf.json --aot $@
```
Comment (tiborm) [inline metron-interface/metron-config/scripts/start-server-for-e2e.sh]: For some reason, git recognized this change as a rename. It is actually a removal of a yml config file (belongs to our previous test server setup, no longer needed) and the addition of a shell script we used to fire up the UI locally.

File: metron-interface/scripts/run-all-ui-test.sh [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+echo "     _/_/_/                          _/_/  _/                _/    _/  _/_/_/   "
+echo "  _/          _/_/    _/_/_/      _/            _/_/_/      _/    _/    _/      "
+echo " _/        _/    _/  _/    _/  _/_/_/_/  _/  _/    _/      _/    _/    _/       "
+echo "_/        _/    _/  _/    _/    _/      _/  _/    _/      _/    _/    _/        "
+echo " _/_/_/    _/_/    _/    _/    _/      _/    _/_/_/        _/_/    _/_/_/       "
+echo "                                                _/                              "
+echo "                                           _/_/                                 "
+echo ""
+
+./metron-config/scripts/run-all-test.sh
```
Comment (tiborm) [inline metron-interface/scripts/run-all-ui-test.sh]: It is hard to distinguish the output of the two UI, therefore, I added the echos above.

File: pom.xml [hunk 11]
```diff
@@ -333,26 +333,26 @@
...
-                        <exclude>e2e/*.js.map</exclude>
+                        <exclude>**/dist/3rdpartylicenses.txt</exclude>
+                        <exclude>**/e2e/*.js.map</exclude>
                         <!-- Checkstyle is LGPL.  We derive ours from their base, but don't ship it, so it's fine use.
                              Since it is derived, we shouldn't get an Apache header -->
                         <exclude>checkstyle.xml</exclude>
                         <!-- Alerts UI e2e test data -->
-                        <exclude>e2e/mock-data/alerts_ui_e2e_index.data</exclude>
+                        <exclude>**/e2e/mock-data/alerts_ui_e2e_index.data</exclude>
                         <!-- retry files generated by Ansible during a failure -->
```
Comment (tiborm) [inline pom.xml]: Without wildcards, the rat check failed to me on my local machine. With this change, it works both local and on Travis.

**Comment by sardell** [comment]
> I also had to remove maven-frontend-plugin to fix issues we faced with on Travis while we tried to run UI test by mvn.

@tiborm Could you give a brief explanation of your research and testing that led to this change?

**Comment by tiborm** [comment]
@sardell Our build jobs started to fail at some point. The cause of the failure was that mvn install was unable to execute unzipping in Cypress post install script without a mysterious error. Few other guys experienced similar [issues](https://github.com/cypress-io/cypress/issues/3515).
I found no other way to fix failing mvn install step other than removing maven-frontend-plugin and migrating to bash scripts. I intentionally left a branch which contains all changes (tests, simplified test server, npm script normalization, removal of extra dependencies, node/npm version bump etc.) and still failing. Because nvm install step failing all other build job fails too most of the time.
U can check it here: https://travis-ci.org/tiborm/metron/builds/500373545
I found no reason for using maven-frontend-plugin.
- It creates a very uncommon execution environment to our build and our tests (node, npm installed by it an also all the following npm scripts). I doubt that any npm module developer tests their install/build process against having this plugin around it.
- It adds extra complexity to the UI projects compared to node version manager supported by Travis natively (even in a language: java setup).
- I also experienced that it slows down the test execution on my local machine.

I found cypress reacted on the issue at the time I writeing this. But not yet proven fix out there.

**Comment by tiborm** [comment]
Please also consider that the Trevis issue was experienced first when added cypress to Config UI. Having it only in a single project (Alert UI in this case) show no issues.
Please also note that the problem is not with Cypress it self. The issue caused by the extracition/unzipping of the Cypress binary made by a module called yauzl. And only if cypress added to multiple projects. This is why I beleive that the issue is somhow related to the dependency cahcing in the mvn install process or the execution context of our cypress install in some way.
The version of the Cypress in Config UI in this branch is exactly the same like we have in Alert UI in the master. Comparision included the version of all the dependecies too.

**Comment by sardell** [comment]
@tiborm Thanks for the explanation. Besides needing this change to resolve the unzipping issue, I also agree that we are adding a layer of complexity by running our front-end tests using maven. I don't see any obvious benefit to doing things this way, and it makes our use case unnecessarily unique. It also makes sense to include this change with this PR since, as you stated, the issue does not happen until you add Cypress to a second UI.

  I’m a tentative +1 for this change, but I would like to first hear what others have to say regarding this change in the PR.

**Comment by tiborm** [comment]
@sardell You are right. I investigated all the possible ways to extract cypress tests from the maven loop but leave the build, unit testing as it is. But no matter what we do 'npm ci' command (like you pointed out) will be executed by maven and that's going to fail if we add cypress to more than one UI project.

For now, I should close this. And hoping for some fix in the future from maven or yauzl.

**Pull request closed without merging**